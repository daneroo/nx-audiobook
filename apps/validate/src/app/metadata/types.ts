export interface AudioBookMetadata {
  author: string
  title: string
  duration: number
  cover?: CoverImageDescriptor
  warning: {
    duration?: string
    cover?: string
  }
}

export interface CoverImageDescriptor {
  size: number
  format: CoverImageFormat
}
export interface CoverImage {
  data: Buffer
  format: CoverImageFormat
}

export type CoverImageFormat = 'image/jpeg' | 'image/png'
