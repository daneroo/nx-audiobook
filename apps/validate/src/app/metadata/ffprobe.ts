import type { FileInfo } from '@nx-audiobook/file-walk'
import { cachedFetchResult } from '../cache/cache'
import type { AudioBookMetadata } from './types'

export async function ffprobe(fileInfo: FileInfo): Promise<AudioBookMetadata> {
  return await cachedFetchResult<FileInfo, AudioBookMetadata>(
    fetchResult,
    fileInfo,
    'ffprobe'
  )
}

// get metadata for a single audio file
export async function fetchResult(
  fileInfo: FileInfo
): Promise<AudioBookMetadata> {
  // const ffprobeBin =
  //   '/Applications/OpenAudible.app/Contents/Resources/app/bin/mac/ffprobe' // ffprobe version 4.3.1
  const ffprobeBin = 'ffprobe'
  const command = `${ffprobeBin} -v quiet -of json -show_format -show_chapters "${fileInfo.path}"`

  try {
    const { stdout, stderr } = await execCommand(command)
    if (stderr !== '') {
      console.error('discarding ffprobe stderr:\n', stderr)
    }
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const metadata = JSON.parse(stdout)
    // console.log(JSON.stringify(metadata, null, 2))
    return {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
      author: metadata.format?.tags?.artist ?? '',
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
      title: metadata.format?.tags?.album ?? '',
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
      duration: Number(metadata.format.duration), // duration is a string, but is it always present?
    }
  } catch (error) {
    console.error({ error })
    throw new Error(`ffprobe error: ${fileInfo.path}`)
  }
}

// Execute a command in a shell
// https://nodejs.org/api/child_process.html#child_processexeccommand-options-callback
//  Good examples for tests
// const results = { example: await execCommand('echo this is stdout') }
// const results = { example: await execCommand('echo this is stderr 1>&2') }
// const results = { example: await execCommand('uname -a') }
// const results = { example: await execCommand('false') } // exit code 1
// const results = {
//   example: await execCommand('docker run --rm ubuntu uname -a')
// }
export async function execCommand(
  command: string
): Promise<{ stdout: string; stderr: string }> {
  const { exec: execWithCallback } = await import('node:child_process')
  const { promisify } = await import('node:util')
  // The promisified version of exec
  const exec = promisify(execWithCallback)
  return await exec(command)
}
