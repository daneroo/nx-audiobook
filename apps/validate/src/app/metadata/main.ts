export { AudioBookMetadata } from './types'
// specific exports partly because of clash with multiple fetchResult
export { getMetadataForSingleFile } from './music-metadata.js'
export { ffprobe } from './ffprobe.js'
