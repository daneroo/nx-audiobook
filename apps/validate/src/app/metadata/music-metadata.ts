import { parseFile } from 'music-metadata'
import type { FileInfo } from '@nx-audiobook/file-walk'
import { cachedFetchResult } from '../cache/cache'
import type { AudioBookMetadata, CoverImage } from './types'

interface MetaOptions {
  duration: boolean
  includeChapters: boolean
}
interface FetchArgs {
  fileInfo: FileInfo
  options: MetaOptions
}

export async function getMetadataForSingleFile(
  fileInfo: FileInfo,
  options: MetaOptions = {
    duration: false, // much slower when true even for some .mp3
    includeChapters: false,
  }
): Promise<AudioBookMetadata> {
  return await cachedFetchResult<FetchArgs, AudioBookMetadata>(
    fetchResult,
    {
      fileInfo,
      options,
    },
    'music-metadata'
  )
}

// get metadata for a single audio file
export async function fetchResult(args: FetchArgs): Promise<AudioBookMetadata> {
  const { fileInfo, options } = args
  try {
    // DO NOT REUSE options object; it gets polluted! {...options} is a workaround
    const metadata = await parseFile(fileInfo.path, { ...options })
    // special case: duration can be NaN, which would turn into null after JSON.stringify|parse
    const duration = isNaN(metadata.format.duration ?? 0)
      ? 0
      : metadata.format.duration ?? 0

    const picture = metadata.common.picture?.[0]

    return {
      author: metadata.common.artist ?? '',
      title: metadata.common.album ?? '',
      duration,
      ...(picture === undefined
        ? {}
        : {
            cover: {
              size: picture.data.length,
              format: picture.format,
            },
          }),
    }
  } catch (error) {
    throw new Error(`music-metadata error: ${fileInfo.path}`)
  }
}

export async function getCoverImage(
  fileInfo: FileInfo
): Promise<CoverImage | undefined> {
  try {
    const metadata = await parseFile(fileInfo.path, { duration: false })
    const pictures = metadata.common.picture
    if (pictures === undefined) {
      return
    }
    if (pictures.length === 0) {
      console.error(`no pictures found in ${fileInfo.path}`)
      return
    }
    if (pictures.length > 1) {
      console.warn(`multiple pictures found in ${fileInfo.path}`)
    }
    const picture = pictures[0]
    if (picture === undefined) {
      console.error(`no pictures[0] found in ${fileInfo.path}`)
      return
    }
    if (!(picture.format === 'image/jpeg' || picture.format === 'image/png')) {
      console.error('cover image is not a jpeg|png file:', picture.format)
      return
    }
    // const cover = metadata.common.picture?.[0].data
    return {
      data: picture.data,
      format: picture.format,
    }
  } catch (error) {
    throw new Error(`music-metadata error: ${fileInfo.path}`)
  }
}
