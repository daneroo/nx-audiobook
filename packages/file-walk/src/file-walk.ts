import { walk, type WalkFunc } from '@root/walk'

import { basename, extname } from 'node:path'
import { promises as fs } from 'node:fs'

// Returns all subdirectories of rootPath, recursively, including rootPath itself
export async function getDirectories(rootPath: string): Promise<string[]> {
  const directories: string[] = []
  // eslint-disable-next-line @typescript-eslint/require-await
  const walker: WalkFunc = async (err, pathname, dirent) => {
    if (err !== null && err !== undefined) {
      // throw an error to stop walking (or return to ignore and keep going)
      console.warn('fs stat error for %s: %s', pathname, err.message)
      return false
    }
    if (dirent.isDirectory()) {
      directories.push(pathname)
      return true
    }
    return false
  }

  await walk(rootPath, walker)
  return directories
}

export interface FileInfo {
  path: string
  basename: string
  extension: string
  size: number
  mtime: Date
}

// Returns all files (as File Objects) from rootPath
// if options.recurse is false, only get files in rootPath, not subdirectories
// if options.stat is false, only get path extension and name, not size or mtime
export async function getFiles(
  rootPath: string,
  options: { recurse: boolean; stat: boolean } = { recurse: false, stat: false }
): Promise<FileInfo[]> {
  const files: FileInfo[] = []

  const walker: WalkFunc = async (err, pathname, dirent) => {
    if (err !== null && err !== undefined) {
      // throw an error to stop walking (or return to ignore and keep going)
      console.warn('fs stat error for %s: %s', pathname, err.message)
      return false
    }
    if (dirent.isDirectory()) {
      // We recurse into subdirectories when either
      // - recurse parameter is true
      // - we are at the root
      return options.recurse || rootPath === pathname
    } else if (dirent.isFile()) {
      const { size, mtime } = options.stat
        ? await fs.stat(pathname)
        : { size: 0, mtime: new Date(0) }
      const file: FileInfo = {
        path: pathname,
        basename: basename(pathname),
        extension: extname(pathname),
        size,
        mtime,
      }
      files.push(file)
    }
    return false
  }
  await walk(rootPath, walker)
  return files
}

// Return a single FileInfo object for the directory at pathname
export async function getDirectory(pathname: string): Promise<FileInfo> {
  const { size, mtime } = await fs.stat(pathname)
  const file: FileInfo = {
    path: pathname,
    basename: basename(pathname),
    extension: extname(pathname),
    size,
    mtime,
  }
  return file
}
