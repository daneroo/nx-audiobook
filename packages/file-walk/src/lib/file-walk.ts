export function fileWalk(): string {
  return 'file-walk';
}
