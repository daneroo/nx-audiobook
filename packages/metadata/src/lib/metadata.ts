export function metadata(): string {
  return 'metadata';
}
