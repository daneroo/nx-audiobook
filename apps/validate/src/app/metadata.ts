import { parseFile } from 'music-metadata'
import type { FileInfo } from '@nx-audiobook/file-walk'
import { isAudioFile } from '@nx-audiobook/validators'

export function metadata(): string {
  return 'metadata'
}

export interface AudioBookMetadata {
  path: string
  author?: string
  title?: string
  duration?: number
}

// get metadata for a single audio file
export async function getMetadataForSingleFile(
  fileInfo: FileInfo,
  options: { duration: boolean; includeChapters: boolean } = {
    duration: false, // much slower when true even for some .mp3
    includeChapters: false,
  }
): Promise<AudioBookMetadata | null> {
  if (!isAudioFile(fileInfo)) {
    return null
  }
  try {
    const metadata = await parseFile(fileInfo.path, options)
    return {
      path: fileInfo.path,
      author: metadata.common.artist ?? '',
      title: metadata.common.album ?? '',
      duration: metadata.format.duration ?? 0,
    }
  } catch (error) {
    console.error(`Error parsing metadata for ${fileInfo.path}`)
    return null
  }
}
