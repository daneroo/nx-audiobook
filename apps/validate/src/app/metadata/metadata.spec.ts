import { describe, expect, it, beforeEach } from 'vitest'
import * as fs from 'node:fs/promises'
import * as path from 'node:path'
import { getFiles } from '@nx-audiobook/file-walk'
import { isAudioFile } from '@nx-audiobook/validators'

import { AudioBookMetadata, getMetadataForSingleFile } from './metadata'

const assetBaseDirectory = path.resolve(
  __dirname,
  '../../../../..',
  'assets',
  'audio'
)

describe('audio assets', () => {
  it('should find assets directory', async () => {
    expect(
      (await fs.readdir(assetBaseDirectory)).filter((f) => f !== '.DS_Store')
    ).toMatchInlineSnapshot(`
      [
        "Derek Walcott - Love After Love",
        "Robert Frost - The Road not Taken",
      ]
    `)
  })
  it('should find audio files', async () => {
    expect(
      (await getFiles(assetBaseDirectory, { recurse: true, stat: false }))
        .map((fi) => fi.basename)
        .filter((basename) => basename !== '.DS_Store')
    ).toMatchInlineSnapshot(`
      [
        "Derek Walcott - Love After Love.mp3",
        "Robert Frost - The Road not Taken.mp3",
      ]
    `)
  })
})

const cacheBaseDirectory = path.join(__dirname, 'testcache')
beforeEach(async () => {
  try {
    await fs.rm(cacheBaseDirectory, { recursive: true })
    process.env.CACHE_BASE_DIRECTORY = cacheBaseDirectory
  } catch (error) {}

  // clean up function, called once after each test run
  return async () => {
    try {
      await fs.rm(cacheBaseDirectory, { recursive: true })
      delete process.env.CACHE_BASE_DIRECTORY
    } catch (error) {}
  }
})

describe('happy path', () => {
  it('should find audio file metadata', async () => {
    const fileInfos = (
      await getFiles(assetBaseDirectory, { recurse: true, stat: true })
    ).filter(isAudioFile)
    expect(fileInfos).toHaveLength(2)

    // run the test twice, second time will hit the cache
    for (let iteration = 0; iteration < 2; iteration++) {
      const metas: AudioBookMetadata[] = []
      for (const fileInfo of fileInfos.filter(isAudioFile)) {
        metas.push(await getMetadataForSingleFile(fileInfo))
      }
      expect(metas).toMatchInlineSnapshot(`
      [
        {
          "author": "Jon Kabat-Zinn, American Public Media",
          "duration": 44.85224489795918,
          "title": "Speaking of Faith from American Public Media",
        },
        {
          "author": "Robert Frost",
          "duration": 76.19918367346939,
          "title": "Poem of the Day from Poetry Foundation.org",
        },
      ]
    `)
    }
  })
})
