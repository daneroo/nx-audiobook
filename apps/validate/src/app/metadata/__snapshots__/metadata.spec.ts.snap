// Vitest Snapshot v1

exports[`happy path > should find audio file metadata (no cache) 1`] = `
[
  {
    "author": "Jon Kabat-Zinn, American Public Media",
    "duration": 44.85224489795918,
    "title": "Speaking of Faith from American Public Media",
  },
  {
    "author": "Robert Frost",
    "duration": 76.19918367346939,
    "title": "Poem of the Day from Poetry Foundation.org",
  },
]
`;

exports[`happy path > should find audio file metadata 1`] = `
[
  {
    "author": "Jon Kabat-Zinn, American Public Media",
    "duration": 44.85224489795918,
    "title": "Speaking of Faith from American Public Media",
  },
  {
    "author": "Robert Frost",
    "duration": 76.19918367346939,
    "title": "Poem of the Day from Poetry Foundation.org",
  },
]
`;
